numba_dpex/core/types/range_types.py

```diff
@@ -0,0 +1,267 @@
+        range_struct.dim1 = range_attr_native_value_map["dim1"].value
+        range_struct.dim2 = range_attr_native_value_map["dim2"].value
+
+    return NativeValue(
+        range_struct._getvalue(), is_error=c.builder.load(is_error_ptr)
+    )
+
+
+@unbox(NdRangeType)
+def unbox_ndrange(typ, obj, c):
+    """
+    Converts a Python Range object to numba-dpex's native struct representation
+    for NdRangeType.
+    """
+    is_error_ptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
+    ndrange_struct = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+
+    with ExitStack() as stack:
+        ndrange_attr_native_value_map = {
+            "global_range": None,
+            "local_range": None,
+        }
+
+        for attr in ndrange_attr_native_value_map.keys():
+            attr_obj = c.pyapi.object_getattr_string(obj, attr)
+            with cgutils.early_exit_if_null(c.builder, stack, attr_obj):
+                c.builder.store(cgutils.true_bit, is_error_ptr)
+            attr_native = c.unbox(RangeType(typ.ndim), attr_obj)
+            c.pyapi.decref(attr_obj)
+            with cgutils.early_exit_if(c.builder, stack, attr_native.is_error):
+                c.builder.store(cgutils.true_bit, is_error_ptr)
+            ndrange_attr_native_value_map[attr] = attr_native
+
+        global_range_struct = ndrange_attr_native_value_map[
+            "global_range"
+        ].value
+        local_range_struct = ndrange_attr_native_value_map["local_range"].value
+
+        range_datamodel = default_manager.lookup(RangeType(typ.ndim))
+        ndrange_struct.ndim = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("ndim"),
+        )
+        ndrange_struct.gdim0 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim0"),
+        )
+        ndrange_struct.gdim1 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.gdim2 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+        ndrange_struct.ldim0 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim0"),
+        )
+        ndrange_struct.ldim1 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.ldim2 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+
+    return NativeValue(
+        ndrange_struct._getvalue(), is_error=c.builder.load(is_error_ptr)
+    )
+
+
+@box(RangeType)
+def box_range(typ, val, c):
+    """
+    Convert a native range structure to a Range object.
+    """
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
```

Comment:
Am I right that this one creates reference on the created object, that's why we don't have to increase reference manually?

Reply:
`ret_ptr` is the `PyObject*` into which the `res` (`Range`) object to be stored is stored. Since it is a new object that gets returned from Numba to CPython, a incref is not needed.

Reply:
When a PyObject is created the ref count should be `1`. Once Numba returns the object to CPython, then CPython will track the lifetime of the object, and decref once the object is out of scope and dead. The garbage collector will free it up at that point.